File: extensions/reddit/src/main/java/app/morphe/extension/reddit/patches/HideAskButtonPatch.java
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches
 *
 * See the included NOTICE file for GPLv3 Section 7 terms that apply to this code.
 */

package app.morphe.extension.reddit.patches;

import app.morphe.extension.reddit.settings.Settings;
import app.morphe.extension.shared.Logger;
import app.morphe.extension.shared.Utils;

@SuppressWarnings("unused")
public final class HideAskButtonPatch {
    private static final String ANDROID_SEARCH_BAR_ASK_BUTTON = "android_search_bar_ask_button";

    /**
     * @return If this patch was included during patching.
     */
    public static boolean isPatchIncluded() {
        return false;  // Modified during patching.
    }


    private static boolean isContextNotYetSet() {
        // Possible fix for background crash of app when context is not yet set.
        if (Utils.getContext() == null) {
            Logger.printInfo(() -> "Cannot hide ask button, context is null");
            return true;
        }
        return false;
    }

    /**
     * Injection point.
     */
    public static boolean hideAskButton(String experimentName, boolean original) {
        // Possible fix for background crash of app when context is not yet set.
        if (isContextNotYetSet()) {
            return original;
        }
        if (Settings.HIDE_ASK_BUTTON.get() && experimentName != null
                && experimentName.startsWith(ANDROID_SEARCH_BAR_ASK_BUTTON)) {
            return false;
        }

        return original;
    }

    /**
     * Injection point.
     */
    public static boolean shouldHideAskButton() {
        if (isContextNotYetSet()) {
            return false;
        }
        return Settings.HIDE_ASK_BUTTON.get();
    }
}

File: patches/src/main/kotlin/app/morphe/patches/reddit/layout/ask/HideAskButtonPatch.kt
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches
 *
 * See the included NOTICE file for GPLv3 §7(b) and §7(c) terms that apply to this code.
 */
package app.morphe.patches.reddit.layout.ask

import app.morphe.patcher.extensions.InstructionExtensions.addInstructionsWithLabels
import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patches.reddit.misc.flag.featureFlagHookPatch
import app.morphe.patches.reddit.misc.flag.hookFeatureFlag
import app.morphe.patches.reddit.misc.settings.settingsPatch
import app.morphe.patches.reddit.misc.version.versionCheckPatch
import app.morphe.patches.reddit.shared.Constants.COMPATIBILITY_REDDIT
import app.morphe.util.setExtensionIsPatchIncluded

private const val EXTENSION_CLASS =
    "Lapp/morphe/extension/reddit/patches/HideAskButtonPatch;"

@Suppress("unused")
val hideAskButtonPatch = bytecodePatch(
    name = "Hide Ask button",
    description = "Adds an option to hide Ask button in the search bar."
) {
    compatibleWith(COMPATIBILITY_REDDIT)

    dependsOn(
        settingsPatch,
        featureFlagHookPatch,
        versionCheckPatch
    )

    execute {

        hookFeatureFlag("$EXTENSION_CLASS->hideAskButton")

        AskButtonComposableFingerprint.method.addInstructionsWithLabels(
            0,
            """
                invoke-static { }, $EXTENSION_CLASS->shouldHideAskButton()Z
                move-result v0
                if-eqz v0, :ignore
                return-void
                :ignore
                nop
            """
        )

        setExtensionIsPatchIncluded(EXTENSION_CLASS)
    }
}
